File: scala/src/main/scala/sodium/StreamWithSend.scala
package sodium

import scala.collection.mutable

class StreamWithSend[A] extends Stream[A] {

  def send(trans: Transaction, a: A): Unit = {
    if (firings.isEmpty) {
      trans.last(() => {
        firings.clear()
      })
    }
    firings += a

    var listeners = mutable.HashSet[Node.Target]()
    Transaction.listenersLock.synchronized {
      listeners = mutable.HashSet() ++ node.listeners
    }

    for (target <- listeners) {
      trans.prioritized(
        target.node,
        trans2 => {
          Transaction.inCallback += 1
          try {
            // Don't allow transactions to interfere with Sodium internals.
            // Dereference the weak reference
            val uta = target.action.get
            if (uta.isDefined) {
              // If it hasn't been gc'ed..., call it
              uta.get match {
                case t: TransactionHandler[_] => t.asInstanceOf[TransactionHandler[A]].run(trans2, a)
                case _                        =>
              }
            }
          } catch {
            case t: Throwable =>
              t.printStackTrace()
          } finally {
            Transaction.inCallback -= 1
          }
        }
      )
    }
  }
}

File: java/sodium/src/sodium/StreamWithSend.java
package sodium;

import java.util.HashSet;

public class StreamWithSend<A> extends Stream<A> {

	protected void send(Transaction trans, A a) {
		if (firings.isEmpty())
			trans.last(new Runnable() {
				public void run() {
					firings.clear();
				}
			});
		firings.add(a);

		HashSet<Node.Target> listeners;
        synchronized (Transaction.listenersLock) {
            listeners = new HashSet<Node.Target>(node.listeners);
        }
		for (Node.Target target : node.listeners) {
            trans.prioritized(target.node, new Handler<Transaction>() {
                public void run(Transaction trans2) {
                    Transaction.inCallback++;
                    try {  // Don't allow transactions to interfere with Sodium
                           // internals.
                        // Dereference the weak reference
                        TransactionHandler<Unit> uta = target.action.get();
                        if (uta != null)  // If it hasn't been gc'ed..., call it
                            ((TransactionHandler<A>)uta).run(trans, a);
                    } catch (Throwable t) {
                        t.printStackTrace();
                    }
                    finally {
                        Transaction.inCallback--;
                    }
                }
            });
		}
	}

}
